tests/torch/external_configurables_test.py

```diff
@@ -0,0 +1,132 @@
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import torch
+
+from absl.testing import absltest
+
+from gin import config
+from gin.torch import external_configurables  # pylint: disable=unused-import
+
+
+@config.configurable
+def fake_train_model(optimizer, scheduler=None):
+    # pylint: disable=E1101
+    opt = optimizer([torch.rand(10), torch.rand(10)])
+    # pylint: enable=E1101
+    sch = None
+    if scheduler:
+        sch = scheduler(opt)
+    return opt, sch
+
+
+@config.configurable
+def configurable(**kwargs):
+    return kwargs
+
+
+class PyTorchConfigTest(absltest.TestCase):
+
+    def assertAlmostEqualList(self, xs, ys):
+        for x, y in zip(xs, ys):
+            self.assertAlmostEqual(x, y)
+
+    def tearDown(self):
+        config.clear_config()
+        super(PyTorchConfigTest, self).tearDown()
+
+    def testConfigureOptimizerAndLearningRate(self):
+        config_str = """
+            fake_train_model.optimizer = @Adam
+            torch.optim.Adam.lr = 0.001
+            torch.optim.Adam.betas = (0.8, 0.888)
+            fake_train_model.scheduler = @StepLR
+            StepLR.step_size = 10
+        """
+        config.parse_config(config_str)
+
+        opt, sch = fake_train_model()  # pylint: disable=no-value-for-parameter
+
+        self.assertIsInstance(opt, torch.optim.Adam)
+        self.assertAlmostEqual(opt.param_groups[0]['betas'][0], 0.8)
+        self.assertAlmostEqual(opt.param_groups[0]['betas'][1], 0.888)
+        self.assertAlmostEqual(opt.defaults['betas'][0], 0.8)
+        self.assertAlmostEqual(opt.defaults['betas'][1], 0.888)
+        self.assertAlmostEqual(sch.step_size, 10)
+
+        lrs = []
+        for _ in range(15):
+            lrs.append(opt.param_groups[0]['lr'])
+            opt.step()
+            sch.step()
+
+        # Divide lr in tenth epoch by 10
+        target_lrs = [0.001, 0.001, 0.001, 0.001, 0.001,
+                      0.001, 0.001, 0.001, 0.001, 0.001,
+                      0.0001, 0.0001, 0.0001, 0.0001, 0.0001]
+
+        self.assertAlmostEqualList(lrs, target_lrs)
+
+    def testOptimizersWithDefaults(self):
+        optimizers = [
+            torch.optim.Adadelta,
+            torch.optim.Adagrad,
+            torch.optim.Adam,
+            torch.optim.SparseAdam,
+            torch.optim.Adamax,
+            torch.optim.ASGD,
+            torch.optim.LBFGS,
+            torch.optim.RMSprop,
+            torch.optim.Rprop,
+            torch.optim.SGD,
+        ]
+        for optimizer in optimizers:
+            config.clear_config()
+            config_lines = ["""fake_train_model.optimizer = @{optimizer}
+                            {optimizer}.lr = 0.001"""
+                            .format(optimizer=optimizer.__name__)]
+            config.parse_config(config_lines)
+
+        # pylint: disable=no-value-for-parameter
```

Comment:
This should be inside the loop, since I suppose you want to test that all the configs are valid optimizers.

Reply:
True, I updated it.

Made me also realize that I used a different indentation level than you guys. Updated that as well.